### PR TITLE
Create deepedit tutorial

### PR DESCRIPTION
**Pull request by diazandr3s**
Signed-off-by: Andres Diaz-Pinto <diazandr3s@gmail.com>

### Description

Add tutorial to run deepedit model on MONAI Core

### Status

Ready

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @diazandr3s ,

Thanks very much for the tutorial!
As we are going to release `MONAI bundle` soon, would you like to also contribute a bundle for this `deepedit` model?
The get_started tutorial and reference example are at: https://github.com/Project-MONAI/tutorials/tree/master/modules/bundles.
And related documentation: https://docs.monai.io/en/latest/bundle_intro.html.

What do you think?
CC @ericspod @wyli 

Thanks in advance.

**Comment by ericspod** [comment]
Thanks for the tutorial!

It's great to have a tutorial introducing DeepEdit but I think it needs more explanation about what's going on and what the moving parts are. The notebook doesn't make clear what the specific elements of DeepEdit are and how they're used, in the first part there is a sequence of transforms which appears to add a landmark into the data dictionary, but the parts which do this aren't explained and the use for this isn't made clear. In the second part a pretrained model is used to segment but how that relates to the first part isn't clear. I understand what the moving pieces are in DeepEdit from other presentations but this needs to be motivated here. 

There's a number of functions in the notebook for plotting, you might want to put these into a PR in core to be kept [here](https://github.com/Project-MONAI/MONAI/blob/dev/monai/visualize/utils.py) if they're generic enough. A general routine for overlaying landmarks would be useful in general.

There's also a lot of code in the `custom_*.py` files, for a tutorial this makes it hard to follow and there seems to be duplication here with what's in MONAILabel itself, could this be reduced or is the point to avoid importing MONAILabel? The notebook imports a number of classes from here that aren't used, it would be clearer if there were no unnecessary imports. 

If would be great if this model were packaged into a bundle with an added inference.json file at least, a training.json file would be great as well that replicated what's in the training script. It would be easier to take the model into other contexts such as deployment.

**Comment by diazandr3s** [comment]
Many thanks for your comments and suggestions, @Nic-Ma @ericspod. They are helpful! :)

I'll create a couple of PRs to MONAI core to include custom transforms, custom interaction, and common plotting functions ... All this to simplify the tutorial. I'll also make the steps/content clearer

I really like the idea of creating a bundle for DeepEdit. I'll work on this once we have the latest version of the tutorial.

Will keep you posted!

**Comment by daguangxu** [comment]
> Many thanks for your comments and suggestions, @Nic-Ma @ericspod. They are helpful! :)
> 
> I'll create a couple of PRs to MONAI core to include custom transforms, custom interaction, and common plotting functions ... All this to simplify the tutorial. I'll also make the steps/content clearer
> 
> I really like the idea of creating a bundle for DeepEdit. I'll work on this once we have the latest version of the tutorial.
> 
> Will keep you posted!

@diazandr3s, we plan to release MONAI 0.9 in mid-June. Can you try to finish this DeepEdit bundle before end of May? Thanks.

**Comment by diazandr3s** [comment]
> > Many thanks for your comments and suggestions, @Nic-Ma @ericspod. They are helpful! :)
> > I'll create a couple of PRs to MONAI core to include custom transforms, custom interaction, and common plotting functions ... All this to simplify the tutorial. I'll also make the steps/content clearer
> > I really like the idea of creating a bundle for DeepEdit. I'll work on this once we have the latest version of the tutorial.
> > Will keep you posted!
> 
> @diazandr3s, we plan to release MONAI 0.9 in mid-June. Can you try to finish this DeepEdit bundle before end of May? Thanks.

Yes, I'll be working on this next week.

**Comment by diazandr3s** [comment]
Quick update on this:

- Add all DeepEdit Transforms and Interaction class in MONAI Core. Here is the PR https://github.com/Project-MONAI/MONAI/pull/4164
- Update the imports in Notebook and train file
- Update the DeepEdit schema and add more information about the way it works

Once I have the MONAI release candidate or MONAI weekly version, I'll update the files

File: deepedit/ignite/train.py [hunk 1]
```diff
@@ -0,0 +1,460 @@
+import argparse
```
Comment (Nic-Ma) [inline deepedit/ignite/train.py:1]: @wyli Do we require to add the copyright for python code?

Thanks.

File: deepedit/ignite/train.py [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+            argmax=(True, False),
+            to_onehot=(len(labels), len(labels)),
+        ),
+        # This transform is to check dice score per segment/label
+        SplitPredsLabeld(keys="pred"),
+    ]
+    return Compose(t)
+
+
+def get_loaders(args, pre_transforms, train=True):
```
Comment (Nic-Ma) [inline deepedit/ignite/train.py:147]: I don't see the usage of arg `train=False` in this code file, maybe review this arg directly?

Thanks.

**Review by Nic-Ma — COMMENTED** [review]
Have you tested both the DynUNet and UNETR? May I know the difference in the result?

Thanks.

**Comment by diazandr3s** [comment]
> Have you tested both the DynUNet and UNETR? May I know the difference in the result?
> 
> Thanks.

Yes! It works better than the DynUNET. But I wanted to keep the DynUNET as default as it takes less time training

File: deepedit/ignite/train.py (hunk 1 above)
Comment (diazandr3s) [reply deepedit/ignite/train.py:1]: Just added the copyright

File: deepedit/ignite/train.py (hunk 2 above)
Comment (diazandr3s) [reply deepedit/ignite/train.py:147]: Agree, thanks!

**Comment by diazandr3s** [comment]
> View / edit / reply to [this conversation](https://app.reviewnb.com/Project-MONAI/tutorials/pull/661/discussion/) on ReviewNB
> 
> _Nic-Ma commented on 2022-05-09T09:55:05Z_ ----------------------------------------------------------------
> 
> Please add these optional imports in above cell directly.
> 
> Refer to other tutorial:
> 
> https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d_lightning.ipynb
> 
> Thanks.

Agree. Just changed it

**Comment by Nic-Ma** [comment]
> > View / edit / reply to [this conversation](https://app.reviewnb.com/Project-MONAI/tutorials/pull/661/discussion/) on ReviewNB
> > _Nic-Ma commented on 2022-05-09T09:55:06Z_ ----------------------------------------------------------------
> > Line #35.        print("Final PLOT:: {} => image shape: {}, pred shape: {}; min: {}, max: {}, sum: {}".format(
> > I think here it plot too many charts, could you please help reduce them? Maybe less than 10 charts?
> > Thanks.
> 
> Agree!

I think you don't need to add another variable `num_slides`, just `for i in range(10)` or `for i in range(0, 100, 10)`?

image

Thanks.

**Comment by diazandr3s** [comment]
> > > View / edit / reply to [this conversation](https://app.reviewnb.com/Project-MONAI/tutorials/pull/661/discussion/) on ReviewNB
> > > _Nic-Ma commented on 2022-05-09T09:55:06Z_ ----------------------------------------------------------------
> > > Line #35.        print("Final PLOT:: {} => image shape: {}, pred shape: {}; min: {}, max: {}, sum: {}".format(
> > > I think here it plot too many charts, could you please help reduce them? Maybe less than 10 charts?
> > > Thanks.
> > 
> > 
> > Agree!
> 
> I think you don't need to add another variable `num_slides`, just `for i in range(10)` or `for i in range(0, 100, 10)`?
> 
> image
> 
> Thanks.

agree!

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me.

**Review by Nic-Ma — COMMENTED** [review]
Looks good to me.

Thanks.
